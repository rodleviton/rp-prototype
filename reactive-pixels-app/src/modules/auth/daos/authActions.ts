import { IAuthUserModel } from "reactive-pixels-common/models/AuthUserModel";
import { createAsyncAction, createStandardAction } from "typesafe-actions";

const PREFIX = "@@auth";

// ---------------- Like Auth User Pixels ------------------ //
/**
 * Note: this action is only triggered via proxy through the rootTransducer.
 * This avoids coupling actions between independent modules.
 */
interface IUpdateAuthUserLikedPixels {
  user: {
    id: string;
    likedPixels: string[];
  };
}

export const updateAuthUserLikedPixels = createStandardAction(
  `${PREFIX}/UPDATE_AUTH_USER_LIKED_PIXELS`
)<IUpdateAuthUserLikedPixels>();

// ---------------- INITIALISE SOCIAL AUTH ---------------- //

export interface IInitAuthUserSuccess {
  authToken: string;
  githubUserId: string;
  id: string;
  refreshToken: string;
}

// These action specifically cover the social auth provider flow
// Github...
// Once we have the initial social auth data we
// can then go and fetch the auth user
export const initAuthUser = createAsyncAction(
  `${PREFIX}/INIT_AUTH_USER_REQUEST`,
  `${PREFIX}/INIT_AUTH_USER_SUCCESS`,
  `${PREFIX}/INIT_AUTH_USER_FAILURE`
)<void, IInitAuthUserSuccess, Error>();

// ---------------- Fetch Authenticated User ---------------- //

export interface IFetchAuthUserSuccess {
  readonly user: IAuthUserModel;
}

export const fetchAuthUser = createAsyncAction(
  `${PREFIX}/FETCH_AUTH_USER_REQUEST`,
  `${PREFIX}/FETCH_AUTH_USER_SUCCESS`,
  `${PREFIX}/FETCH_AUTH_USER_FAILURE`
)<void, IFetchAuthUserSuccess, Error>();
