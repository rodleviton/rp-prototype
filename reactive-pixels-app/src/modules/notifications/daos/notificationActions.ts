// import { INotificationModel } from "@modules/notifications/models/NotificationModel";
import { createStandardAction } from "typesafe-actions";

const PREFIX = "@@notification";

export const showNotification = createStandardAction(
  `${PREFIX}/SHOW_NOTIFICATION`
)<string>();

export const hideNotification = createStandardAction(
  `${PREFIX}/HIDE_NOTIFICATION`
)();
