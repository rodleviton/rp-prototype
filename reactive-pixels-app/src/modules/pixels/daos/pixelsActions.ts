import { ApiMethods } from "@modules/common/api/apiConstants";
import { IPixelsCommentModel } from "reactive-pixels-common/models/PixelsCommentModel";
import { IPixelsMetadataModel } from "reactive-pixels-common/models/PixelsMetadataModel";
import { IPixelsModel } from "reactive-pixels-common/models/PixelsModel";
import { createAsyncAction, createStandardAction } from "typesafe-actions";

// ---------------- Like Pixels ------------------ //
interface ILikePixelsRequest {
  pixelsId: string;
  userId: string;
  method: ApiMethods.Delete | ApiMethods.Post;
}

const PREFIX = "@@pixels";

export const likePixels = createAsyncAction(
  `${PREFIX}/LIKE_PIXELS_REQUEST`,
  `${PREFIX}/LIKE_PIXELS_SUCCESS`,
  `${PREFIX}/LIKE_PIXELS_FAILURE`
)<ILikePixelsRequest, void, Error>();

// ---------------- Fetch Pixels ---------------- //
interface IFetchPixelsRequest {
  id?: string;
  owner?: string;
}

export interface IFetchPixelsSuccess {
  readonly allIds: ReadonlyArray<string>;
  readonly itemsById: { [id: string]: IPixelsModel };
}

export const fetchPixels = createAsyncAction(
  `${PREFIX}/FETCH_PIXELS_REQUEST`,
  `${PREFIX}/FETCH_PIXELS_SUCCESS`,
  `${PREFIX}/FETCH_PIXELS_FAILURE`
)<IFetchPixelsRequest, IFetchPixelsSuccess, Error>();

// ---------------- Select Pixels ---------------- //

export const selectPixels = createStandardAction(`${PREFIX}/SELECT_PIXELS`)<
  string
>();

// ---------------- Fetch Pixels Meta ---------------- //
interface IFetchPixelsMetaRequest {
  id?: string;
}

export interface IFetchPixelsMetaSuccess {
  readonly meta: IPixelsMetadataModel;
}

export const fetchPixelsMeta = createAsyncAction(
  `${PREFIX}/FETCH_PIXELS_META_REQUEST`,
  `${PREFIX}/FETCH_PIXELS_META_SUCCESS`,
  `${PREFIX}/FETCH_PIXELS_META_FAILURE`
)<IFetchPixelsMetaRequest, IFetchPixelsMetaSuccess, Error>();

// ---------------- Fetch Pixels Comments ---------------- //
interface IFetchPixelsCommentsRequest {
  id?: string;
}

export interface IFetchPixelsCommentsSuccess {
  readonly comments: IPixelsCommentModel[];
}

export const fetchPixelsComments = createAsyncAction(
  `${PREFIX}/FETCH_PIXELS_COMMENTS_REQUEST`,
  `${PREFIX}/FETCH_PIXELS_COMMENTS_SUCCESS`,
  `${PREFIX}/FETCH_PIXELS_COMMENTS_FAILURE`
)<IFetchPixelsCommentsRequest, IFetchPixelsCommentsSuccess, Error>();
