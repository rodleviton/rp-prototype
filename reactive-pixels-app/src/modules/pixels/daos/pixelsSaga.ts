import { ApiMethods } from "@modules/common/api/apiConstants";
import { get, request } from "@modules/common/api/async";
import {
  fetchPixels,
  fetchPixelsComments,
  fetchPixelsMeta,
  likePixels
} from "@modules/pixels/daos/pixelsActions";
import {
  buildFetchPixelsCommentsUrl,
  buildFetchPixelsMetaUrl,
  buildFetchPixelsUrl,
  buildLikePixelsUrl
} from "@modules/pixels/helpers/urlBuilder";
import { pixelsDataToPixelsTransformer } from "@modules/pixels/mappers/PixelsDataToPixelsTransformerFactory";
import { call, put, takeEvery } from "redux-saga/effects";
import { ActionType, getType } from "typesafe-actions";

// ---------------- Fetch Pixels ---------------- //

export async function fetchPixelsFromAPI(pixelsId?: string, owner?: string) {
  const url = buildFetchPixelsUrl(pixelsId, owner);

  return get({ url });
}

function* fetchPixelsRequest(params: ActionType<typeof fetchPixels.request>) {
  const { id, owner } = params.payload;
  const [error, pixels] = yield call(fetchPixelsFromAPI, id, owner);

  if (pixels) {
    let pixelsData;

    // Run some data type checks
    if (pixels && Array.isArray(pixels.data) && pixels.data.length > 0) {
      pixelsData = pixels.data;
    } else if (pixels && !Array.isArray(pixels.data) && pixels.data) {
      pixelsData = [pixels.data];
    } else {
      pixelsData = [];
    }

    yield put({
      payload: pixelsDataToPixelsTransformer.convertMultiple(pixelsData),
      type: getType(fetchPixels.success)
    });
  }

  if (error) {
    yield put({ type: getType(fetchPixels.failure) });
  }
}

// ---------------- Like Pixels ------------------ //

async function likePixelsFromAPI(pixelsId: string, method: ApiMethods) {
  const url = buildLikePixelsUrl(pixelsId);

  return request({ method, url });
}

function* likePixelsRequest(params: ActionType<typeof likePixels.request>) {
  const { method, pixelsId } = params.payload;
  const [error, result] = yield call(likePixelsFromAPI, pixelsId, method);

  if (result) {
    yield put({
      payload: result.data,
      type: getType(likePixels.success)
    });
  }

  if (error) {
    yield put({ type: getType(likePixels.failure) });
  }
}

// ---------------- Fetch Pixels Meta ---------------- //

export async function fetchPixelsMetaFromAPI(pixelsId: string) {
  const url = buildFetchPixelsMetaUrl(pixelsId);

  return get({ url });
}

function* fetchPixelsMetaRequest(
  params: ActionType<typeof fetchPixelsMeta.request>
) {
  const { id } = params.payload;
  const [error, pixelsMeta] = yield call(fetchPixelsMetaFromAPI, id);

  if (pixelsMeta) {
    yield put({
      payload: pixelsMeta.data,
      type: getType(fetchPixelsMeta.success)
    });
  }

  if (error) {
    yield put({ type: getType(fetchPixelsMeta.failure) });
  }
}

// ---------------- Fetch Pixels Comments ---------------- //

export async function fetchPixelsCommentsFromAPI(pixelsId: string) {
  const url = buildFetchPixelsCommentsUrl(pixelsId);

  return get({ url });
}

function* fetchPixelsCommentsRequest(
  params: ActionType<typeof fetchPixelsComments.request>
) {
  const { id } = params.payload;
  const [error, pixelsComments] = yield call(fetchPixelsCommentsFromAPI, id);

  if (pixelsComments) {
    yield put({
      payload: pixelsComments.data,
      type: getType(fetchPixelsComments.success)
    });
  }

  if (error) {
    yield put({ type: getType(fetchPixelsComments.failure) });
  }
}

export default function* fetchPixelsSaga() {
  yield takeEvery(getType(fetchPixels.request), fetchPixelsRequest);
  yield takeEvery(getType(likePixels.request), likePixelsRequest);
  yield takeEvery(getType(fetchPixelsMeta.request), fetchPixelsMetaRequest);
  yield takeEvery(
    getType(fetchPixelsComments.request),
    fetchPixelsCommentsRequest
  );
}
