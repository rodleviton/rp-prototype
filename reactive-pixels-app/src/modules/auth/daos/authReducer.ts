import * as authActions from "@modules/auth/daos/authActions";
import { IAuthUserModel } from "reactive-pixels-common/models/AuthUserModel";
import { combineReducers } from "redux";
import { ActionType, getType } from "typesafe-actions";

export type AuthAction = ActionType<typeof authActions>;

const { updateAuthUserLikedPixels, fetchAuthUser } = authActions;

export interface IAuthState {
  readonly user: IAuthUserModel;
}

export const initialState: IAuthState = {
  user: {} as IAuthUserModel
};

export default combineReducers<IAuthState>({
  user: (state = initialState.user, action) => {
    switch (action.type) {
      case getType(fetchAuthUser.request):
        return {}; // clear out existing data

      case getType(fetchAuthUser.success):
        return action.payload;

      case getType(fetchAuthUser.failure):
        return state;

      // ---------------- Like Pixels ------------------ //

      case getType(updateAuthUserLikedPixels): {
        const { id, likedPixels } = action.payload.user;

        if (state.id === id) {
          return {
            ...state,
            likedPixels
          };
        }

        return state;
      }
      default:
        return state;
    }
  }
});
