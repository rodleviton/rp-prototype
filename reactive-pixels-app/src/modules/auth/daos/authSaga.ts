import { fetchAuthUser, initAuthUser } from "@modules/auth/daos/authActions";
import {
  buildCreatAuthUserUrl,
  buildFetchAuthUserUrl
} from "@modules/auth/helpers/urlBuilder";
import { get, post } from "@modules/common/api/async";
import { call, put, takeEvery } from "redux-saga/effects";
import { ActionType, getType } from "typesafe-actions";

export async function fetchAuthenticatedUserFromAPI(id: string) {
  const url = buildFetchAuthUserUrl(id);

  return get({ url });
}

export async function createAuthenticatedUserFromAPI(id: string) {
  const url = buildCreatAuthUserUrl(id);

  // We have all the information in the request headers to create/update a user
  return post({ url, data: null });
}

function* fetchAuthUserRequest(
  params: ActionType<typeof initAuthUser.success>
) {
  yield put({ type: getType(fetchAuthUser.request) });

  const { id, authToken, refreshToken, githubUserId } = params.payload;

  // TODO - add all User details to AuthUser??
  const [error, authUser] = yield call(createAuthenticatedUserFromAPI, id);

  if (authUser) {
    yield put({
      payload: {
        authToken,
        githubUserId,
        refreshToken,
        ...authUser.data
      },
      type: getType(fetchAuthUser.success)
    });
  }

  if (error) {
    yield put({ type: getType(fetchAuthUser.success) });
  }
}

export default function* fetchAuthenticatedUserSaga() {
  yield takeEvery(getType(initAuthUser.success), fetchAuthUserRequest);
}
