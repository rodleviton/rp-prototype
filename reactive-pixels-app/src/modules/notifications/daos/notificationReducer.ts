import * as notificationActions from "@modules//notifications/daos/notificationActions";
import { ActionType, getType } from "typesafe-actions";

export type NotificationAction = ActionType<typeof notificationActions>;

const { hideNotification, showNotification } = notificationActions;

export interface INotificationsState {
  modalType: notificationActions.Notifications;
}

export const initialState: INotificationsState = {
  modalType: ""
};

export default (state = initialState, action: any) => {
  switch (action.type) {
    case getType(showNotification):
      return { modalType: action.payload };

    case getType(hideNotification):
      return {}; // clear out existing data

    default:
      return state;
  }
};
