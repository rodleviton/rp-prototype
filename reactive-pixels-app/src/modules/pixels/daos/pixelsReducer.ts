import { ApiMethods } from "@modules/common/api/apiConstants";
import {
  addItemToArray,
  removeItemFromArray
} from "@modules/common/utils/arrayUtilities";
import * as pixelsActions from "@modules/pixels/daos/pixelsActions";
import { IPixelsCommentModel } from "reactive-pixels-common/models/PixelsCommentModel";
import { IPixelsMetadataModel } from "reactive-pixels-common/models/PixelsMetadataModel";
import { IPixelsModel } from "reactive-pixels-common/models/PixelsModel";
import { combineReducers } from "redux";
import { ActionType, getType } from "typesafe-actions";

export type PixelsAction = ActionType<typeof pixelsActions>;

const {
  fetchPixels,
  fetchPixelsComments,
  fetchPixelsMeta,
  likePixels,
  selectPixels
} = pixelsActions;

export interface IPixelsState {
  readonly allIds: ReadonlyArray<string>;
  readonly itemsById: { [id: string]: IPixelsModel };
  readonly selected: {
    comments: IPixelsCommentModel[];
    id: string;
    meta: IPixelsMetadataModel | {};
  };
}

export const initialState: IPixelsState = {
  allIds: [],
  itemsById: {},
  selected: {
    comments: [],
    id: "",
    meta: {}
  }
};

export default combineReducers<IPixelsState>({
  allIds: (state = initialState.allIds, action) => {
    switch (action.type) {
      case getType(fetchPixels.request):
        return []; // clear out existing data

      case getType(fetchPixels.success):
        return action.payload.allIds;

      case getType(fetchPixels.failure):
        return state;

      default:
        return state;
    }
  },
  itemsById: (state = initialState.itemsById, action) => {
    // ---------------- Fetch Pixels ---------------- //

    switch (action.type) {
      case getType(fetchPixels.request):
        return {}; // clear out existing data

      case getType(fetchPixels.success):
        return action.payload.itemsById;

      case getType(fetchPixels.failure):
        return state;

      // ---------------- Like Pixels ------------------ //

      case getType(likePixels.request): {
        const { method, pixelsId, userId } = action.payload;
        // Here we optimistically update the state
        // TODO - UPDATE STATE ON SUCCESS/FAILURE to ensure state is always in sync
        const updatedLikes =
          method === ApiMethods.Delete
            ? removeItemFromArray(state[pixelsId].likes, userId)
            : addItemToArray(state[pixelsId].likes, userId);

        return {
          ...state,
          [pixelsId]: {
            ...state[pixelsId],
            likes: updatedLikes
          }
        };
      }

      default:
        return state;
    }
  },
  selected: (state = initialState.selected, action) => {
    switch (action.type) {
      // select
      case getType(selectPixels):
        return {
          comments: [],
          id: action.payload,
          meta: {}
        };

      // comments
      case getType(fetchPixelsComments.success):
        return {
          comments: action.payload,
          id: state.id,
          meta: state.meta
        };

      // meta
      case getType(fetchPixelsMeta.success):
        return {
          comments: state.comments,
          id: state.id,
          meta: action.payload
        };

      default:
        return state;
    }
  }
});
