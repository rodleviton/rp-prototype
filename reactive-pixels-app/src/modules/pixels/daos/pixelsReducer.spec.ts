import {
  pixels1,
  pixelsStateFixture
} from "@modules/pixels/__fixtures__/pixelsFixtures";
import * as actions from "@modules/pixels/daos/pixelsActions";
import reducer, { initialState } from "@modules/pixels/daos/pixelsReducer";

/**
 * STORIES
 */
describe("Pixels Stories", () => {
  describe("initial state", () => {
    it("should match a snapshot", () => {
      expect(reducer(initialState, {} as any)).toMatchSnapshot();
    });
  });

  describe("fetching pixels", () => {
    describe("fetch pixels request", () => {
      it("should `reset` state", () => {
        const action = actions.fetchPixels.request({
          id: "123456"
        });
        const state = reducer(pixelsStateFixture, action);

        expect(state.allIds).toHaveLength(0);
        expect(state.itemsById).toEqual({});
      });
    });

    describe("fetch pixels success", () => {
      it("should `populate` state", () => {
        const action = actions.fetchPixels.success(pixelsStateFixture);
        const state = reducer(initialState, action);

        expect(state.allIds).toHaveLength(1);
        expect(state.allIds[0]).toEqual(pixels1.id);
        expect(state.itemsById).toEqual({
          [pixels1.id]: pixels1
        });
      });
    });

    describe("fetch pixels failure", () => {
      it("should `return` state", () => {
        const action = actions.fetchPixels.failure(Error("Oops."));
        const state = reducer(initialState, action);

        expect(state.allIds).toEqual(initialState.allIds);
        expect(state.itemsById).toEqual(initialState.itemsById);
      });
    });
  });

  describe("selecting pixels", () => {
    it("should set `selected` pixels", () => {
      const action = actions.selectPixels("l7G5RpHUMt4vGslDyz9q");
      const state = reducer(initialState, action);

      expect(state.selected).toEqual("l7G5RpHUMt4vGslDyz9q");
    });
  });
});
