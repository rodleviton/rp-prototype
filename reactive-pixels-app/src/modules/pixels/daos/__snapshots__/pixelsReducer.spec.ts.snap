// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pixels Stories initial state should match a snapshot 1`] = `
Object {
  "allIds": Array [],
  "itemsById": Object {},
  "selected": "",
}
`;
